instance-scheduler/main.go

```diff
@@ -0,0 +1,273 @@
+			}
+
+			actedUponMessage := fmt.Sprintf("%v instance %v (ReservationId: %v) %v\n", action, *i.InstanceId, *r.ReservationId, instanceSchedulingTagDescr)
+			skippedMessage := fmt.Sprintf("Skipped instance %v (ReservationId: %v) %v\n", *i.InstanceId, *r.ReservationId, instanceSchedulingTagDescr)
+			// Tag key: instance-scheduling
+			// Valid values: default (same as absence of tag), skip-scheduling, skip-auto-stop, skip-auto-start
+			if (instanceSchedulingTag == "") || (instanceSchedulingTag == "default") {
+				log.Print(actedUponMessage)
+				instancesActedUpon = append(instancesActedUpon, *i.InstanceId)
+				if action == "Stop" {
+					stopInstance(client, *i.InstanceId)
+				} else {
+					startInstance(client, *i.InstanceId)
+				}
+			} else if instanceSchedulingTag == "skip-scheduling" {
+				log.Print(skippedMessage)
+				skippedInstances = append(skippedInstances, *i.InstanceId)
+			} else if instanceSchedulingTag == "skip-auto-stop" {
+				if action == "Stop" {
+					log.Print(skippedMessage)
+					skippedInstances = append(skippedInstances, *i.InstanceId)
+				} else {
+					log.Print(actedUponMessage)
+					instancesActedUpon = append(instancesActedUpon, *i.InstanceId)
+					startInstance(client, *i.InstanceId)
+				}
+			} else if instanceSchedulingTag == "skip-auto-start" {
+				if action == "Stop" {
+					log.Print(actedUponMessage)
+					instancesActedUpon = append(instancesActedUpon, *i.InstanceId)
+					stopInstance(client, *i.InstanceId)
+				} else {
+					log.Print(skippedMessage)
+					skippedInstances = append(skippedInstances, *i.InstanceId)
+				}
+			}
+		}
+	}
+
+	acted := "Started"
+	if action == "Stop" {
+		acted = "Stopped"
+	}
+	if len(instancesActedUpon) > 0 {
+		log.Printf("%v %v instances: %v\n", acted, len(instancesActedUpon), instancesActedUpon)
+	} else {
+		log.Printf("WARN: No instances were %v!\n", acted)
+	}
+	if len(skippedInstances) > 0 {
+		log.Printf("Skipped %v instances: %v\n", len(skippedInstances), skippedInstances)
+	}
+}
+
+func getEc2ClientForMemberAccount(cfg aws.Config, accountName string, accountId string) *ec2.Client {
+	roleARN := fmt.Sprintf("arn:aws:iam::%v:role/MemberInfrastructureAccess", accountId)
+	stsClient := sts.NewFromConfig(cfg)
+	provider := stscreds.NewAssumeRoleProvider(stsClient, roleARN)
+	cfg.Credentials = aws.NewCredentialsCache(provider)
+
+	iamClient := iam.NewFromConfig(cfg)
+	_, err := iamClient.ListRoles(context.TODO(), &iam.ListRolesInput{
+		PathPrefix: aws.String("/")})
+
+	if err != nil {
+		if strings.Contains(err.Error(), "is not authorized to perform: sts:AssumeRole on resource") {
+			log.Printf("WARN: account %v (%v) is ignored because it does not have the role MemberInfrastructureAccess, therefore is not a member account\n", accountName, accountId)
+			return nil
+		} else {
+			log.Fatal(err)
+		}
+	}
+	return ec2.NewFromConfig(cfg)
+}
+
+func handler(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
```

Comment:
If we're doing this on a schedule, and not triggering it via an API call, I think this might need `event events.CloudWatchEvent` rather than `events.APIGatewayProxyRequest`